### ISSUE TITLE
Enable a read-only mode for Kinto where the database user can work with read-only permissions

### ISSUE THREAD

**Issue by Natim** [Kinto/kinto#235]

**Comment by tarekziade** [comment]
The simplest and fastest way to do it is to create a wsgi middleware that rejects all calls where request.method != 'GET'

**Comment by Natim** [comment]
This should already be possible with http://kinto.readthedocs.org/en/latest/configuration/settings.html#enabling-or-disabling-endpoints

Also I am wondering if a new setting ``kinto.read-only = true`` could help to configure that behavior.

**Comment by leplatrem** [comment]
As said somewhere else :

- The most efficient way to make sure that Kinto does write to the database is to configure it with a PostgreSQL user that has only readonly privileges
- The inconvenient is that every requests PUT/POST/PATCH/DELETE will be responded with a 503, will trigger Sentry errors.

We have thus to limit the requests to GET.
This can happen at several levels:
- Reverse proxy (or CDN) - ie. Nginx
- At the software level (during view registration)
- At the wsgi middleware level (in kinto init)

The higher the better. So I would be in favor of doing it in Nginx + PostgreSQL of course. But as we say, we can take «a belt and braces approach» and do something in the Kinto codebase :)

With the middleware approach, we have to be careful because this will restrist the batch requests. Even though every subrequest is readonly.

With the view registration approach (as suggested by Remy), we are sure that it applies to REST resources views only.

With a `kinto.read_only_mode` setting, we can make sure that the list of disabled resources (buckets, collections, groups, records) is exhaustive...

My 2 cents...

**Comment by tarekziade** [comment]
> So I would be in favor of doing it in Nginx + PostgreSQL of course

Ops are in favour of the app level because it simplifies their puppets deployments if they just create a  custom read-only kinto config.

In the final deployement they will block any write calls at the CDN level, but they want a safety net.

> With the middleware approach, we have to be careful because this will restrist the batch requests. Even though every subrequest is readonly.

I guess this can be detected in the middleware by introspecting the request body ? 
This is also very important to know for the CDN rules. 

bonus question is : are we using this POST-is-a-bunch-of-GET feature for the sec settings ?

**Comment by leplatrem** [comment]
> I guess this can be detected in the middleware by introspecting the request body ? 

Yes I guess we can just introspect the path (`request.path.startswith('/batch'))

> this POST-is-a-bunch-of-GET feature

Nope. This is a very specific usage of the batch endpoint, for which we don't have any use-case currently. (There used to be one for readinglist, where the UI would want to populate several menu entries by article read status using only one request)

So, for sec settings, at the CDN level we can go for the easiest way : only GET.

But in Kinto, in order to remain consistent, I suggest we allow POST on batch. (or we only disable resource verbs using available settings, as suggested by Remy)

**Comment by leplatrem** [comment]
After quick conversation with @tarekziade, the approach suggested by @Natim is the way to go. There is no code to write, just manipulate setttings at startup when readonly is set, and the views won't be registered in Pyramid (405 responses when accessed)

**Comment by Natim** [comment]
So I tried my investigation. The aim I have is to safely let people configure a read-only postgresql user on a read-only kinto.

I did a ``cliquet migrate`` with the postgres rw user

Then the first error I have doing that when running the server is:

```
ERROR:venusian:(psycopg2.ProgrammingError) ERREUR:  doit être le propriétaire de la fonction sec2ttl
 [SQL: "--\n-- Automated script, we do not need NOTICE and WARNING\n--\nSET client_min_messages TO ERROR;\n\nCREATE TABLE IF NOT EXISTS cache (\n    key VARCHAR(256) PRIMARY KEY,\n    value TEXT NOT NULL,\n    ttl TIMESTAMP DEFAULT NULL\n);\n\n--\n-- CREATE INDEX IF NOT EXISTS will be available in PostgreSQL 9.5\n-- http://www.postgresql.org/docs/9.5/static/sql-createindex.html\nDO $$\nBEGIN\n\n  IF NOT EXISTS (\n    SELECT 1 FROM pg_indexes\n       WHERE indexname = 'idx_cache_ttl'\n       AND tablename = 'cache'\n  ) THEN\n\n  CREATE INDEX idx_cache_ttl ON cache(ttl);\n\n  END IF;\nEND$$;\n\n\nCREATE OR REPLACE FUNCTION sec2ttl(seconds FLOAT)\nRETURNS TIMESTAMP AS $$\nBEGIN\n    IF seconds IS NULL THEN\n        RETURN NULL;\n    END IF;\n    RETURN now() + (seconds || ' SECOND')::INTERVAL;\nEND;\n$$ LANGUAGE plpgsql;\n"] 
Traceback (most recent call last):
  File ".venv/bin/cliquet", line 9, in <module>
    load_entry_point('cliquet==2.10.0.dev0', 'console_scripts', 'cliquet')()
  File "/home/rhubscher/mozilla/kinto/.venv/local/lib/python2.7/site-packages/cliquet/scripts/cliquet.py", line 46, in main
    args.func(env)
  File "/home/rhubscher/mozilla/kinto/.venv/local/lib/python2.7/site-packages/cliquet/scripts/cliquet.py", line 21, in init_schema
    getattr(registry, backend).initialize_schema()
  File "/home/rhubscher/mozilla/kinto/.venv/local/lib/python2.7/site-packages/cliquet/cache/postgresql/__init__.py", line 66, in initialize_schema
    conn.execute(schema)
  File "/usr/lib/python2.7/contextlib.py", line 35, in __exit__
    self.gen.throw(type, value, traceback)
  File "/home/rhubscher/mozilla/kinto/.venv/local/lib/python2.7/site-packages/cliquet/storage/postgresql/client.py", line 51, in connect
    raise exceptions.BackendError(original=e)
cliquet.storage.exceptions.BackendError: ProgrammingError: (psycopg2.ProgrammingError) ERREUR:  doit être le propriétaire de la fonction sec2ttl
 [SQL: "--\n-- Automated script, we do not need NOTICE and WARNING\n--\nSET client_min_messages TO ERROR;\n\nCREATE TABLE IF NOT EXISTS cache (\n    key VARCHAR(256) PRIMARY KEY,\n    value TEXT NOT NULL,\n    ttl TIMESTAMP DEFAULT NULL\n);\n\n--\n-- CREATE INDEX IF NOT EXISTS will be available in PostgreSQL 9.5\n-- http://www.postgresql.org/docs/9.5/static/sql-createindex.html\nDO $$\nBEGIN\n\n  IF NOT EXISTS (\n    SELECT 1 FROM pg_indexes\n       WHERE indexname = 'idx_cache_ttl'\n       AND tablename = 'cache'\n  ) THEN\n\n  CREATE INDEX idx_cache_ttl ON cache(ttl);\n\n  END IF;\nEND$$;\n\n\nCREATE OR REPLACE FUNCTION sec2ttl(seconds FLOAT)\nRETURNS TIMESTAMP AS $$\nBEGIN\n    IF seconds IS NULL THEN\n        RETURN NULL;\n    END IF;\n    RETURN now() + (seconds || ' SECOND')::INTERVAL;\nEND;\n$$ LANGUAGE plpgsql;\n"]
```

**Comment by leplatrem** [comment]
Of course, you won't touch the schema with a readonly user :) `cliquet migrate` is irrelevant here no ?

**Comment by Natim** [comment]
Yes it is because ``make serve`` automatically run ``make migrate`` first.

**Comment by Natim** [comment]
A get works

**Comment by Natim** [comment]
Fixed in #525
